### DISCUSSION TITLE
Keytar Migration Plan

### DISCUSSION THREAD

**Discussion by TylerLeonhardt** [microsoft/vscode-discussions#662] [Announcements]
Hello :wave: I wanted to give you all a heads up around our plan for dealing with our dependency on [Keytar](https://github.com/atom/node-keytar), the abstraction layer we have been using to store secrets in the OS keyring... along with the old Keytar shim that extensions depended on before the introduction of the [SecretStorage API](https://code.visualstudio.com/api/references/vscode-api#SecretStorage).

Keytar itself is [officially archived](https://github.com/atom/node-keytar) and not being maintained... In an effort to promote good security practices by _not_ depending on an archived piece of software for storing secrets, we are working on a plan to remove this shim from VS Code.

We have already directly reached out to extension authors of popular extensions that still use this Keytar shim... and they are working on moving off of it. We want to communicate this here as well for those we didn't contact directly so we minimize the disruption due to this change. I know this isn't the best news... we had a fair amount of work to remove our dependency on Keytar as well, but we believe this is the right thing to do to ensure extensions are using secure APIs.

If you are using Keytar directly in your extension, there are a couple of options for you to consider:
* _(recommended)_ Use the [SecretStorage API](https://code.visualstudio.com/api/references/vscode-api#SecretStorage) that VS Code provides on the [ExtensionContext](https://code.visualstudio.com/api/references/vscode-api#ExtensionContext). This API is cross-platform and will work on all platforms that VS Code supports. It is also maintained by the VS Code team and will continue to be maintained and has been a part of the VS Code API for years at this point.
* _(not recommended)_ You can bundle the [keytar](https://github.com/atom/node-keytar) module with your extension. Keep in mind that keytar is a native node module which means that you will need to [publish a platform specific extension](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions) for each platform you want to support.

### Timeline

The current plan is to remove this shim from VS Code Insiders in **early September** which means that **September 2023's** stable release will be the first release without the shim.

Work for this is being tracked here: https://github.com/microsoft/vscode/issues/185677

### Questions?

If you have any questions, please feel free to ask them here. I will do my best to answer them as quickly as possible. Your cooperation is greatly appreciated :heart:

**Comment by wilhelmklopp** [comment]
👋 Hi @TylerLeonhardt – We're running into an issue when testing our VSCode extension, which I suspect is related to this change. More details here: https://github.com/microsoft/vscode/issues/188059

**Comment by adam-wolfe** [comment]
Just to confirm, by "September 2023's stable release", do you mean the release that is scheduled for release in early October?

**Comment by TylerLeonhardt** [comment]
@adam-wolfe the timeline was pushed back a month... but yes that is a correct interpretation of what I said. So now, we will consider removing it in early September which means that September 2023's stable release (often released in the first week of Oct) will be the first release without the shim.

**Reply by adam-wolfe** [reply]
Excellent, thanks.

**Reply by TylerLeonhardt** [reply]
@adam-wolfe I misspoke in this and edited my message. Since we removed it in early September, it will be in September's stable release (on track to be released next week). Apologies for that confusion.

**Reply by adam-wolfe** [reply]
Not a problem, we are ready.

**Comment by Kreijstal** [comment]
Can't build vscode on arm32 because of the keytar dependency

**Reply by TylerLeonhardt** [reply]
Keytar is no longer a dependency in VS Code. It hasn't been in 3+ months. Can you provide more information?
